src/main/java/org/elasticsearch/indices/IndicesService.java

```diff
@@ -419,4 +418,128 @@ public synchronized void beforeIndexShardClosed(ShardId shardId, @Nullable Index
+            logger.debug("[{}] failed to delete index store - at least one shards is still locked", ex, index);
+        } catch (Exception ex) {
+            logger.warn("[{}] failed to delete index", ex, index);
+        } finally {
+            // this is a pure protection to make sure this index doesn't get re-imported as a dangeling index.
+            // we should in the future rather write a tombstone rather than wiping the metadata.
+            MetaDataStateFormat.deleteMetaState(nodeEnv.indexPaths(index));
+        }
+    }
+
+    /**
+     * Deletes the shard with an already acquired shard lock. Note
+     * @param reason the reason for the shard deletion
+     * @param lock the lock of the shard to delete
+     * @param indexSettings the shards index settings.
+     * @throws IOException if an IOException occurs
+     */
+    public void deleteShardStore(String reason, ShardLock lock, Settings indexSettings) throws IOException {
+        ShardId shardId = lock.getShardId();
+        if (canDeleteShardContent(shardId, indexSettings) == false) {
+            throw new ElasticsearchIllegalStateException("Can't delete shard " + shardId);
+        }
+        logger.info("{} deleting shard reason [{}]", shardId, reason);
+        nodeEnv.deleteShardDirectoryUnderLock(lock, indexSettings);
+    }
+
+    /**
+     * This method deletes the shard contents on disk for the given shard ID. This method will fail if the shard deleting
+     * is prevented by {@link #canDeleteShardContent(org.elasticsearch.index.shard.ShardId, org.elasticsearch.cluster.metadata.IndexMetaData)}
+     * of if the shards lock can not be acquired.
+     * @param reason the reason for the shard deletion
+     * @param shardId the shards ID to delete
+     * @param metaData the shards index metadata. This is required to access the indexes settings etc.
+     * @throws IOException if an IOException occurs
+     */
+    public void deleteShardStore(String reason, ShardId shardId, IndexMetaData metaData) throws IOException {
+        final Settings indexSettings = buildIndexSettings(metaData);
+        if (canDeleteShardContent(shardId, indexSettings) == false) {
+            throw new ElasticsearchIllegalStateException("Can't delete shard " + shardId);
+        }
+        logger.info("{} deleting shard reason [{}]", shardId, reason);
+        nodeEnv.deleteShardDirectorySafe(shardId, indexSettings);
+    }
+
+    /**
+     * Returns <code>true</code> iff the shards content for the given shard can be deleted.
+     * If the shard is still allocated / active on this node or for instance if the shard is located on shared
+     * filesystem this method will return <code>false</code>.
+     *
+     * @param shardId the shard to delete.
+     * @param metaData the shards index metadata. This is required to access the indexes settings etc.
+     */
+    public boolean canDeleteShardContent(ShardId shardId, IndexMetaData metaData) {
```

Comment:
Also here I wonder if we should be safe and synchronize this. Do we really need the metaData? we can get the setting from the injector (which we check anyway). Also I think a better name is hasShardUsedContent (we return false if there is nothing to delete.